Updated functions, added mock test

Sudhintc opened the pull request:
1.Splitting functions into distinct blocks of code
2.Data sanitisation from user input
3.Error handling using exceptions
4.added mock test

Sudhintc committed:
Revert "added-functionality"

This reverts commit 2849d903d843b8c27bc439c7a4d08c882e33e3f6.

Sudhintc committed:
Revert "added-functionality"

Sudhintc committed:
added mock test

gavin-mannings-wl commented on index.test.js:
```diff
@@ -0,0 +1,71 @@
+
+describe("getWeather", () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
```
You can wrap you `describe` functions so that you only need one `jest.resetAllMocks()` call:

```javascript
describe('Weather app:', () => {

  beforeEach(() => {
      jest.resetAllMocks();
  });

  describe("getWeather", () => {
    ...
  });

  describe("api error", () => {
   ...
  });

})

gavin-mannings-wl commented on index.test.js:
```diff
@@ -0,0 +1,71 @@
+  beforeEach(() => {
+    jest.resetAllMocks();
+  });
+  test("should fetch weather data from API and display it", async () => {
```
Try to make `test` methods only perform a test against one block of code. Here there are two, I shall write up an example of re-organizing your tests that may make this easier.

gavin-mannings-wl commented on index.test.js:
```diff
@@ -0,0 +1,71 @@
+    `;
+
+    await getWeather();
+    displayWeatherData(mockedData);
```
Is there a reason to call this directly here rather than allow the code to progress and call it as part of the `getWeather` function? I can see on line 14 you'd expect this to be called here.

gavin-mannings-wl commented on index.test.js:
```diff
@@ -0,0 +1,71 @@
+    const invalidCity = "not valid";
+    global.fetch = jest
+      .fn()
+      .mockRejectedValue(new Error("weather data not found"));
```
This is only testing the code from line 16-22, should there be a test that covers the block 9-10?

gavin-mannings-wl commented:
@Sudhintc This is a good unit test introduction, well done!

I was confused as your PR mentioned a few things that weren't shown as differences in the code. Did you force push some changes to this repo? I've added comments to https://github.com/Sudhintc/js-weather-app/commit/2849d903d843b8c27bc439c7a4d08c882e33e3f6 that I think cover off these areas, it's a good first pass!

For organizing your unit tests it can be helpful to nest describe blocks (each describe doesn't need an assertion), this also nests the test results in the result window when you run the test. This grouping can make sharing functionality easier and clearer, and also makes test reports easier to read.

For example, organizing the tests like this:

```javascript

describe('Weather app:', () => {

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('getWeather()', () => {

    describe('should fetch weather data from API', () => {

      describe('when there is a successful API response', () => {

        beforeEach(() => {
          global.fetch = jest.fn().mockResolvedValue(mockedData);
        });

        test('it should display the weather', ...);

      });

      describe('when there is an unsuccessful API response', () => {

        beforeEach(() => {
          new Error("weather data not found");
          global.fetch = jest.fn().mockRejectedValue(error);
        });

        test('it should display the error', ...);

      });

    });

  });

});
```

Will provide an output like:

```
Weather app:
  getWeather()
    should fetch weather data from API
      when there is a successful API response
        it should display the weather
      when there is an unsuccessful API response
        it should display the error
```

Which is a useful way to see which test is failing when methods and functions become more complex.

Sudhintc commented:
@gavin-mannings-wl Thank you for your feedback and taking time to review my pull request.
 When committing the new code I accidently made it to the main branch , so I tried to revert the commit , and I think that's why you couldn't see the changes .
also , I will make the suggested changes to the code 

Thank You

Sudhintc committed:
updated-suggestions

Sudhintc commented:
updated test.js file as suggested
Added showWeather ()
completed all other suggestion except hiding api key and sanitizing city
tried api_key hiding using env file but cannot use require in browser environment , and also was not able to use import as well

gavin-mannings-wl commented:
Thanks Sudhin, apologies for how long its taken me to get back to you.

The changes all look very good!

For the .env/API key the idea is to prevent users from looking at your JavaScript and stealing your API key. You cannot achieve this without having a proxy between the API and your application as any client-side code is always accessible to a user. A small NodeJS/Express, PHP, or other server-side endpoint to proxy your API request is the standard solution. This will allow the server to add the API key without it being exposed on the client-side. A .env file can be used to hold the API key for the server-side app to load in at runtime.

Regarding sanitization of the city name, we generally want to handle XSS and invalid characters that will cause an error. This can be handled on the client-side or on the server-side, ideally both.

The following function would cover the basics of sanitization:

```javascript
function sanitizeInput(input) {
  // Using the String.prototype.replace() method with a regular expression 
  // to remove any character that isn't a letter (unicode), and then 
  // uses another regular expression to remove HTML tags
  const regexNonLetters = /[^\p{L}]/gu;
  const regexHtml = /<[^>]*>?/gm;

  let sanitizedInput = input.replace(regexNonLetters, '');
  sanitizedInput = sanitizedInput.replace(regexHtml, '');

  return sanitizedInput;
}
```

This uses regular expressions to replace non-text unicode characters that should work for most use cases. https://www.regular-expressions.info/unicode.html#prop has a lot of details on using regular expressions to handle unicode.

Ideally we would use a library for security sanitization.

gavin-mannings-wl reviewed (approved):
Looks very good!
